LLVM errors since upgrading from llvm 13 to 14

I package crystal for Solus, and since updating to LLVM 14 we get the following errors. Previously, with LLVM 13, all compiler specs passed and the interpreter worked. Both I and the LLVM maintainer are at a loss. I don't even know for sure if crystal is at fault here but no other packages have issues since the upgrade.

The compiler to all intents and purposes works fine (at least it can bootstrap itself), only the tests and interpreter fail. From #11905 I know that LLVM 14.0 _should_ be supported.

```console
$ make -j8 compiler_spec CRYSTAL_CONFIG_TARGET=x86_64-linux-gnu
Using /usr/bin/llvm-config [version= 14.0.6]
CRYSTAL_CONFIG_BUILD_COMMIT="" CRYSTAL_CONFIG_PATH='$ORIGIN/../share/crystal/src' SOURCE_DATE_EPOCH="1665087238" ./bin/crystal build -D strict_multi_assign --release --link-flags="-Wl,--copy-dt-needed-entries -Wl,-O1 -Wl,-z,relro -Wl,-z,now -Wl,-z,max-page-size=0x1000 -Wl,-Bsymbolic-functions -Wl,--sort-common" -Dwithout_interpreter  --exclude-warnings spec/std --exclude-warnings spec/compiler --exclude-warnings spec/primitives -o .build/compiler_spec spec/compiler_spec.cr
/snip
compiler_spec: /home/build/YPKG/root/llvm/build/llvm-project-14.0.6.src/llvm/include/llvm/IR/DataLayout.h:653: uint64_t llvm::StructLayout::getElementOffset(unsigned int) const: Assertion `Idx < NumElements && "Invalid element idx!"' failed.
make: *** [Makefile:92: compiler_spec] Aborted (core dumped)
```

```console
$ crystal i
crystal: /home/build/YPKG/root/llvm/build/llvm-project-14.0.6.src/llvm/include/llvm/Support/Casting.h:269: typename cast_retty<X, Y *>::ret_type llvm::cast(Y *) [X = llvm::StructType, Y = llvm::Type]: Assertion `isa<X>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
Aborted (core dumped)
```

--- a/spec/compiler/codegen/offsetof_spec.cr
+++ b/spec/compiler/codegen/offsetof_spec.cr
@@ -56,4 +56,12 @@ describe "Code gen: offsetof" do
       (pointerof(f).as(Void*) + offsetof(Foo, @y).to_i64).as(UInt32*).value == f.y
       CRYSTAL
   end
+
+  it "returns offset of `StaticArray#@buffer`" do
+    run(<<-CRYSTAL).to_b.should be_true
+      x = uninitialized Int32[4]
+      pointerof(x.@buffer).value = 12345
+      (pointerof(x).as(Void*) + offsetof(Int32[4], @buffer).to_i64).as(Int32*).value == x.@buffer
+      CRYSTAL
+  end
 end

--- a/spec/compiler/interpreter/pointers_spec.cr
+++ b/spec/compiler/interpreter/pointers_spec.cr
@@ -121,6 +121,32 @@ describe Crystal::Repl::Interpreter do
       CRYSTAL
     end
 
+    it "pointerof read `StaticArray#@buffer` (1)" do
+      interpret(<<-CRYSTAL).should eq(2)
+        struct StaticArray(T, N)
+          def to_unsafe
+            pointerof(@buffer)
+          end
+
+          def x
+            @buffer
+          end
+        end
+
+        foo = uninitialized Int32[4]
+        foo.to_unsafe.value = 2
+        foo.x
+        CRYSTAL
+    end
+
+    it "pointerof read `StaticArray#@buffer` (2)" do
+      interpret(<<-CRYSTAL).should eq(2)
+        foo = uninitialized Int32[4]
+        pointerof(foo.@buffer).value = 2
+        foo.@buffer
+        CRYSTAL
+    end
+
     it "interprets pointer set and get (union type)" do
       interpret(<<-CRYSTAL).should eq(10)
         ptr = Pointer(Int32 | Bool).malloc(1_u64)

--- a/src/compiler/crystal/codegen/codegen.cr
+++ b/src/compiler/crystal/codegen/codegen.cr
@@ -105,13 +105,13 @@ module Crystal
     end
 
     def offset_of(type, element_index)
-      return 0_u64 if type.extern_union?
+      return 0_u64 if type.extern_union? || type.is_a?(StaticArrayInstanceType)
       llvm_typer.offset_of(llvm_typer.llvm_type(type), element_index)
     end
 
     def instance_offset_of(type, element_index)
-      # extern unions must be value types, which always use the above
-      # `offset_of` instead
+      # extern unions and static arrays must be value types, which always use
+      # the above `offset_of` instead
       llvm_typer.offset_of(llvm_typer.llvm_struct_type(type), element_index + 1)
     end
   end

